### PR TITLE
fixes #1049 - apoc.export.cypher.all fails when a compound constraint exists

### LINKED ISSUE #1049: apoc.export.cypher.all fails when a compound constraint exists
## Expected Behavior (Mandatory)
`CALL apoc.export.cypher.all('/data/seed.cypher',{format:'cypher-shell'});` works when compound constraint exist.

## Actual Behavior (Mandatory)
```
CALL apoc.export.cypher.all('/data/seed.cypher',{format:'cypher-shell'});
Failed to invoke procedure `apoc.export.cypher.all`: Caused by: java.util.NoSuchElementException: More than one element in java.util.AbstractList$Itr@693ec7a0. First element is 'adTagType' and the second element is 'name'
```

## How to Reproduce the Problem
Create a compound constraint on a node. I have the following:

```
create constraint on (t:AdTag) assert (t.adTagType, t.name) is node key;
```

### Simple Dataset (where it's possibile)

```
MERGE (a:AdTag:test {name: "E", adTagType: "ESRB-test"});
MERGE (a:AdTag:test {name: "Under 18", adTagType: "Age-test"});
MERGE (a:AdTag:test {name: "13-17",  adTagType: "Age-test"});
MERGE (a:AdTag:test {name: "18-24", adTagType: "Age-test"});
MERGE (a:AdTag:test {name: "25-34", adTagType: "Age-test"});
```

### Steps (Mandatory)

  1. Run `create constraint on (t:AdTag) assert (t.adTagType, t.name) is node key;`
  1. Add the sample data above.
  1. Run `CALL apoc.export.cypher.all('/data/seed.cypher',{format:'cypher-shell'});`

## Specifications (Mandatory)

Currently used versions

### Versions

  - OS: OS X 10.13.6
  - Neo4j: Enterprise 3.5.0
  - Neo4j-Apoc: 3.5.0.1

### PR DESCRIPTION
**Pull request by AngeloBusato**
Fixes #1049 

apoc.export.cypher.all fails when a compound constraint exists

## Proposed Changes (Mandatory)

A brief list of proposed changes in order to fix the issue:

  - compound constraind are now managed

### CODE REVIEW

**Comment by jexp** [comment]
For compound keys we also have to create MERGE and MATCh statements with multiple keys to assert the concrete uniqueness

So we might turn the query generator into always using multiple properties and then a size=1 is just a special case.

File: src/main/java/apoc/export/cypher/MultiStatementCypherSubGraphExporter.java [hunk 1]
```diff
@@ -160,7 +160,12 @@ private void exportSchema(PrintWriter out) {
             String label = index.getLabel().name();
             Iterable<String> props = index.getPropertyKeys();
             if (index.isConstraintIndex()) {
-                String cypher = this.cypherFormat.statementForConstraint(label, Iterables.single(props));
+                String cypher;
+                if(Iterables.count(props) > 1) {
+                    cypher = this.cypherFormat.statementForCompoundConstraint(label, props);
```
Comment (jexp) [inline src/main/java/apoc/export/cypher/MultiStatementCypherSubGraphExporter.java]: can't we handle that inside of the this.cypherFormat ?

**Comment by jexp** [comment]
So mean mean this as a fix to the original exception but not as a proper fix b/c those composite keys are actually ignored and lead to different nodes being collapsed on whichever property comes first in an constraint?

**Comment by jexp** [comment]
@conker84 ^^^

**Comment by AngeloBusato** [comment]
@jexp PR Updated

**Comment by jexp** [comment]
@sarmbruster suggested to move the enterprise tests to a testcontainer, at least for 3.5 we'd need that.

What about my comment above? 
Can we make sure that it doesn't happen then?

I.e. you have  two nodes with a composite key, and you export a relationship.
What is used as "MATCH" key?

**Comment by conker84** [comment]
@jexp  I'll handle the #1079 ok?

**Comment by conker84** [comment]
> you have two nodes with a composite key, and you export a relationship.
> What is used as "MATCH" key?

We'll add a test case for that

**Comment by AngeloBusato** [comment]
@jexp PR Updated

**Comment by jexp** [comment]
Thanks a lot.

**Pull request merged**